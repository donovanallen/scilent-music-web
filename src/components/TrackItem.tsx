import { SimplifiedTrack, Track } from '@spotify/web-api-ts-sdk';
import Link from 'next/link';
import React from 'react';
import { FaMusic } from 'react-icons/fa6';
import { GiBackwardTime } from 'react-icons/gi';

import NextImage from '@/components/NextImage';

import {
  cn,
  formatArtists,
  getDurationText,
  getTimestampText,
} from '../lib/utils';

interface TrackItemProps {
  track: Track | SimplifiedTrack;
  timestamp?: Date;
  className?: string;
  numbered?: boolean;
  disabled?: boolean;
}

const TrackItem: React.FC<TrackItemProps> = ({
  track,
  timestamp,
  className,
  numbered = false,
  disabled = false,
}) => {
  const getReleaseId = (track: Track) => {
    return track.id;
  };

  return (
    <Link
      href={!disabled ? `/release/${getReleaseId(track as Track)}` : ''}
      aria-disabled={disabled}
      className={disabled ? 'cursor-default' : 'cursor-pointer'}
    >
      <div
        className={cn(
          'w-full items-center px-4 py-2 flex gap-x-4 h-fit rounded-md overflow-hidden ',
          !disabled
            ? 'hover:bg-neutral-800/40 hover:text-brand-dark transition'
            : 'hover:bg-neutral-800/40 transition',
          className,
        )}
      >
        {numbered ? (
          <div className='mr-2'>
            <h4 className='subtitle'>{track.track_number}</h4>
          </div>
        ) : (
          <div className='relative min-h-[64px] min-w-[64px] rounded-md overflow-hidden bg-neutral-700'>
            {'album' in track ? (
              <NextImage
                src={track.album.images[0].url}
                alt={track.album.name + ' image'}
                fill
                useSkeleton
                className='object-cover aspect-square'
              />
            ) : (
              <FaMusic size={36} className='m-auto h-full text-dark' />
            )}
          </div>
        )}
        <div className='flex flex-col h-full items-start justify-between flex-1'>
          <div className='w-full overflow-clip'>
            <h4 className='font-normal truncate'>{track.name}</h4>
            <p className='subtitle font-normal text-neutral-500 truncate'>
              {formatArtists(track.artists)}
            </p>
          </div>
          {timestamp && (
            <div className='flex gap-x-1 items-center text-neutral-500'>
              <GiBackwardTime />
              <span className='subtitle'>
                {getTimestampText(timestamp.toString())}
              </span>
            </div>
          )}
        </div>
        <div>
          <h4 className='subtitle text-neutral-500'>
            {getDurationText(track.duration_ms)}
          </h4>
        </div>
      </div>
    </Link>
  );
};

export default TrackItem;
