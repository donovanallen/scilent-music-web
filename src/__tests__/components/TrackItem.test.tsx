import { Track } from '@spotify/web-api-ts-sdk';
import { render } from '@testing-library/react';
import React from 'react';

import TrackItem from '@/components/TrackItem';

describe('TrackItem', () => {
  const mockTrack = {} as Track;
  it('renders correctly', () => {
    const { asFragment } = render(<TrackItem track={mockTrack} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
