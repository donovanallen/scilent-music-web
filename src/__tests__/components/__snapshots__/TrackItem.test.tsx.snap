// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrackItem renders correctly 1`] = `
<DocumentFragment>
  <a
    aria-disabled="false"
    class="cursor-pointer"
    href="/release/undefined"
  >
    <div
      class="w-full items-center px-4 py-2 flex gap-x-4 h-fit rounded-md overflow-hidden hover:bg-neutral-800/40 hover:text-brand-dark transition"
    >
      <div
        class="relative min-h-[64px] min-w-[64px] rounded-md overflow-hidden bg-neutral-700"
      >
        <svg
          class="m-auto h-full text-dark"
          fill="currentColor"
          height="36"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="36"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M499.1 6.3c8.1 6 12.9 15.6 12.9 25.7v72V368c0 44.2-43 80-96 80s-96-35.8-96-80s43-80 96-80c11.2 0 22 1.6 32 4.6V147L192 223.8V432c0 44.2-43 80-96 80s-96-35.8-96-80s43-80 96-80c11.2 0 22 1.6 32 4.6V200 128c0-14.1 9.3-26.6 22.8-30.7l320-96c9.7-2.9 20.2-1.1 28.3 5z"
          />
        </svg>
      </div>
      <div
        class="flex flex-col h-full items-start justify-between flex-1"
      >
        <div
          class="w-full overflow-clip"
        >
          <h4
            class="font-normal truncate"
          />
          <p
            class="subtitle font-normal text-neutral-500 truncate"
          />
        </div>
      </div>
      <div>
        <h4
          class="subtitle text-neutral-500"
        >
          NaN:NaN:NaN
        </h4>
      </div>
    </div>
  </a>
</DocumentFragment>
`;
